Render questions using filesystem

Crola1702 opened the pull request:
# Description

Instead of creating a Nanoc DataSource to load all questions from a json dump, we want to use Filesystem data source and render all questions, answers and comments from the question metadata.

# Changes

This PR:
* Removes outdated compilation rules to keep only `/question/*.md` rule.
* Removes `questions` DataSource
* Adds a sample markdown question: [:scroll: 1.md](https://github.com/ros-infrastructure/answers-archive/blob/markdown-questions-dump/content/question/1.md?plain=1).
* Adds a helper function to render markdown from a string.
  > **Note**: This was the way I managed to render answers and comments `BodyMarkdown`, as they're not generated as items, so we can't apply redcarpet filter.
* Update layouts to render answers and comments based in question item metadata

I see answers.gazebosim stores all the questions under `/question/:id` endpoint, I believe it is better to have the same directory structure, so I created `content/question/` also.

Example output of `1.md` [gazebosim-1-html.pdf](https://github.com/user-attachments/files/16380370/gazebosim-1-html.pdf)
> GitHub doesn't allow adding html files

Crola1702 committed:
Render questions using filesystem

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Add explanation comment

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 commented:
@nuclearsandwich I recall talking about compiling two times, one for gazebosim and other for ros? How is the website split idea?

nuclearsandwich commented on content/question/1.md:
```diff
@@ -0,0 +1,103 @@
+Tags: gazebo
+Title: how to stop the joint motion( clear the force)
+UpvoteCount: 0
+---
```
I'm very tempted to ask for metadata to be in in `snake_case` format matching Ruby / Python identifier conventions. If that's immeasurably hard then we'll roll with this.

Crola1702 replied:
Yeah that's doable. Will take a look

Crola1702 replied:
Done in 9bfef93

nuclearsandwich commented on layouts/answer.html:
```diff
@@ -1,22 +1,18 @@
+  <% if item[:Accepted] %>
   <div class="accepted">
-    <span class="accepted">☑️</span> This answer was <strong>ACCEPTED</strong> on the original site.
+    <span class="accepted">☑️ </span> This answer was <strong>ACCEPTED</strong> on the original site.
```
Why is a space needed inside the span with the emoji rather than only outside it?

Crola1702 replied:
Good question. I don't know where did it come from :thinking:

Crola1702 replied:
Removed in 7e1fd06

nuclearsandwich commented on layouts/comment.html:
```diff
@@ -1,6 +1,4 @@
-  <div class="owner">
-    <p>Comment by <%= item[:owner].display_name %> on <%= pretty_date item[:creation_date] %></p>
-  </div>
+  <div class="owner"><p>Asked by <%= item[:OwnerDisplayName] %> on <%= pretty_date item[:CreationDate] %></p></div>
```
TODO: we'll want this to eventually be a link to the author's profile (index of questions asked and answered)

Crola1702 replied:
Should we also generate author pages?

nuclearsandwich commented on content/question/1.md:
```diff
@@ -0,0 +1,103 @@
+---
+Answers:
+- Accepted: true
+  BodyMarkdown: "Where in you'r code have you applied the force ?\r\nare you absolutely\
+    \ sure it is not in the onUpdate() or similar never ending call ?\r\n\r\nWhat\
+    \ is the purpose?\r\n\r\nhave you got friction in the joint ?\r\naccording to\
+    \ newtons laws of motion, the arm will continue to rotate forever, unless something\
+    \ provides a reaction force, like friction.\r\n\r\nif you apply the force in a\
+    \ small duration of time, the turnrate of the arm will after that moment in time\
+    \ become constant. With frictions it will decellerate, with repeated force the\
+    \ arm will accellerate.\r\nwith repeated force and friction the arm will reach\
+    \ a steady state where you'r applied force equals the force from friction, and\
+    \ from here the speed will eventually become constant.\r\n\r\n**solutionset:**\r\
+    \n\r\n1. if you are repeatingly applying force, stop that after some time and\
+    \ let the friction stop the arm (use gazebos simulation time..)\r\n\r\n2. learn\
+    \ control theory, (and basic physics) and let a controller that calculate the\
+    \ desired force to have the arm move to a position or speed or both be implemented.\
+    \  (You should do this to control the HW anyway)\r\n\r\n3. stop using forces,\
+    \ _cheat_ and use the animation interface for gazebo\r\n"
```
This format parses but it isn't the literal scalar format we discussed. Was it not possible to generate scalar literals after all?

Crola1702 replied:
I'm having a weird problem there. Some questions are generated as scalar literals and some are not. I'm still debugging why is this happening

Crola1702 replied:
Just to save every step I take from here:

If the dumped string contains a "\r" character, the string is dumped in quotes (that's my investigation for now, I'm not sure why is that.

Crola1702 replied:
Now it seems everything has the literal now. (Just replaced "\r" for "" in all body markdowns)

nuclearsandwich commented on lib/helpers.rb:
```diff
@@ -47,6 +47,12 @@ def migrated_gz_posts
   end
 end
 
+def render_markdown md_str
+  # This is inefficient as markdown renderer should not be created every time
+  markdown = Redcarpet::Markdown.new(Redcarpet::Render::HTML)
+  markdown.render(md_str)
+end
+
```
I'm pushing a commit that I think helps here.

Crola1702 replied:
Do you think there is a better way to render markdown in metadata? I thought about using a module or class and create only one renderer, but for that it seems I need esoteric ruby.

nuclearsandwich committed:
Move render_markdown helper into a helper module.

This also provides us with a place to store our markdown parser between
function invocations by using a module constant.

nuclearsandwich commented on content/question/1.md:
```diff
@@ -0,0 +1,103 @@
+  OwnerProfileUrl: https://answers.gazebosim.org/users/3828/adaman05/
+  UpvoteCount: 0
+Comments:
+- BodyMarkdown: i have the same kind of problem ,i tried to modify the c++ code
+    in tutorial http://gazebosim.org/wiki/Tutorials/1.3/control_robot/mobile_base_laser
+    .it also not stopping.i have setForce to zero even it is not stooping .I am
+    beginner in field of robotics as well as in gazebo .plz guide me ,i am attachin
+    my modified code here
```
It's weird that this is unqouted and still seems to work fine. I expect because of the indentation.

Crola1702 replied:
That's true. I'll update this post in a new dump example version

Crola1702 replied:
Updated in 9bfef93

nuclearsandwich committed:
Specify redcarpet renderer options.

These match the current behavior of Askbot.

nuclearsandwich commented on content/question/1.md:
```diff
@@ -0,0 +1,103 @@
+  OwnerProfileUrl: https://answers.gazebosim.org/users/3828/adaman05/
+  UpvoteCount: 0
+Comments:
+- BodyMarkdown: i have the same kind of problem ,i tried to modify the c++ code
+    in tutorial http://gazebosim.org/wiki/Tutorials/1.3/control_robot/mobile_base_laser
+    .it also not stopping.i have setForce to zero even it is not stooping .I am
+    beginner in field of robotics as well as in gazebo .plz guide me ,i am attachin
+    my modified code here
```
Is this comment copied here to test the layout? It's not here on the current Answers site: https://answers.gazebosim.org/question/1/how-to-stop-the-joint-motion-clear-the-force/

Crola1702 replied:
Yeah, good catch! I was testing everything worked properly, so I copied and pasted a comment.

Crola1702 replied:
The new question example ins updated now

nuclearsandwich commented:
> I recall talking about compiling two times, one for gazebosim and other for ros? How is the website split idea?

My uncomplicated idea is basically to maintain `gazebo` and `ros` branches of this repository, each with the data for the respective site and cherry-pick shared changes between them.

A more complicated idea is to store no content in the `main` branch and add helper scripts to check `ros` or `gz` content out before a build.

Crola1702 committed:
Refactor everything to snake_case

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Remove emoji whitespace

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Comment rules while we make them work

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

Crola1702 committed:
Questions and authors indexes (#2)

* Update indexes to support only <PLACEHOLDER> site

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add questions index

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Sort questions by last created (tested with 7257 questions from gazebosim

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add author and creation date in question index layout

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Refactor question list to use renderer helper

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add author profile and question index

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add links to authors

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add placeholder in page_title helper

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Add title in nanoc config

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>

* Filter the question index with ERB instead of markdown.

Becuase there is very little content on this page, there's not much
motivation to have markdown for that content and using erb allows us to
put the questions listing directly in the content file instead of adding
a single-purpose layout file to use as a snippet.

---------

Signed-off-by: Crola1702 <cristobal.arroyo@ekumenlabs.com>
Co-authored-by: Steven! Ragnarök <steven@nuclearsandwich.com>
